### ISSUE TITLE
dtest test failure: out of limit commitlog size 

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#12129]
Seen in e.g. https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/149/testReport/commitlog_test/TestCommitLog/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_total_space_limit_of_commitlog_with_memory_based_limit/

Digging into the full test log shows the following:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/149/artifact/logs-full.release.000/dtest-gw1.log
```
05:58:33,190 541     commitlog_test                 DEBUG    commitlog_test.py   :850  | test_total_space_limit_of_commitlog_with_memory_based_limit: commitlog_segment_size_in_mb=85
05:58:33,190 541     commitlog_test                 DEBUG    commitlog_test.py   :851  | test_total_space_limit_of_commitlog_with_memory_based_limit: commitlog_total_space_in_mb=-1
05:58:33,190 541     commitlog_test                 DEBUG    commitlog_test.py   :852  | test_total_space_limit_of_commitlog_with_memory_based_limit: total_space_limit=512
05:58:33,203 541     commitlog_test                 DEBUG    commitlog_test.py   :866  | test_total_space_limit_of_commitlog_with_memory_based_limit: Commitlog size before start: (0, 0)M
05:58:33,203 541     commitlog_test                 DEBUG    commitlog_test.py   :867  | test_total_space_limit_of_commitlog_with_memory_based_limit: Start cluster with `--smp 1' ...
05:58:33,203 541     ccm                            DEBUG    cluster.py          :711  | test_total_space_limit_of_commitlog_with_memory_based_limit: start_nodes: no_wait=False wait_for_binary_proto=True wait_other_notice=True force_wait_for_cluster_start=True
05:58:33,209 541     ccm                            DEBUG    cluster.py          :711  | test_total_space_limit_of_commitlog_with_memory_based_limit: node1: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--smp', '1', '--api-address', '127.0.31.1', '--collectd-hostname', 'bb22af9e148e.node1', '--memory', '1024M', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--developer-mode', 'true', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.31.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
05:58:34,087 541     ccm                            DEBUG    cluster.py          :711  | test_total_space_limit_of_commitlog_with_memory_based_limit: node1: Starting scylla-jmx: args=['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.31.1', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.31.1', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.31.1', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/bin/scylla-jmx-1.0.jar']
05:58:34,399 541     commitlog_test                 DEBUG    commitlog_test.py   :874  | test_total_space_limit_of_commitlog_with_memory_based_limit: Create test keyspace and table
05:58:34,404 541     cassandra.policies             INFO     policies.py         :289  | test_total_space_limit_of_commitlog_with_memory_based_limit: Using datacenter 'datacenter1' for DCAwareRoundRobinPolicy (via host '127.0.31.1:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
05:58:34,430 541     commitlog_test                 DEBUG    commitlog_test.py   :882  | test_total_space_limit_of_commitlog_with_memory_based_limit: Starting CS workload: ['write', 'n=102400', '-rate', 'threads=25', '-col', 'size=FIXED(1000)', '-log', 'interval=100']
05:58:34,431 541     ccm                            WARNING  cluster.py          :717  | test_total_space_limit_of_commitlog_with_memory_based_limit: node1: passing `capture_output` to stress_object() is deprecated
05:58:39,501 541     commitlog_test                 DEBUG    commitlog_test.py   :911  | test_total_space_limit_of_commitlog_with_memory_based_limit: Current commitlog size: 340M (66.41)%, Well-used cases: 1 / 5, Time passed: 5.07s, commitlog_disk_usage_threshold: 256M
05:58:45,708 541     commitlog_test                 DEBUG    commitlog_test.py   :921  | test_total_space_limit_of_commitlog_with_memory_based_limit: Subprocess_Return(stdout='******************** Stress Settings ********************\nCommand:\n  Type: write\n  Count: 102,400\n  No Warmup: false\n  Consistency Level: LOCAL_ONE\n  Serial Consistency Level: SERIAL\n  Target Uncertainty: not applicable\n  Key Size (bytes): 10\n  Counter Increment Distibution: add=fixed(1)\nRate:\n  Auto: false\n  Thread Count: 25\n  OpsPer Sec: 0\nPopulation:\n  Sequence: 1..102400\n  Order: ARBITRARY\n  Wrap: true\nInsert:\n  Revisits: Uniform:  min=1,max=1000000\n  Visits: Fixed:  key=1\n  Row Population Ratio: Ratio: divisor=1.000000;delegate=Fixed:  key=1\n  Batch Type: not batching\nColumns:\n  Max Columns Per Key: 5\n  Column Names: [C0, C1, C2, C3, C4]\n  Comparator: AsciiType\n  Timestamp: null\n  Variable Column Count: false\n  Slice: false\n  Size Distribution: Fixed:  key=1000\n  Count Distribution: Fixed:  key=5\nErrors:\n  Ignore: false\n  Tries: 10\nLog:\n  No Summary: false\n  No Settings: false\n  File: null\n  Interval Millis: 100000\n  Level: NORMAL\nMode:\n  API: JAVA_DRIVER_NATIVE\n  Connection Style: CQL_PREPARED\n  CQL Version: CQL3\n  Protocol Version: V4\n  Username: null\n  Password: null\n  Auth Provide Class: null\n  Max Pending Per Connection: null\n  Connections Per Host: 8\n  Compression: NONE\nNode:\n  Nodes: [127.0.31.1]\n  Is White List: false\n  Datacenter: null\nSchema:\n  Keyspace: keyspace1\n  Replication Strategy: org.apache.cassandra.locator.SimpleStrategy\n  Replication Strategy Options: {replication_factor=1}\n  Table Compression: null\n  Table Compaction Strategy: null\n  Table Compaction Strategy Options: {}\nTransport:\n  factory=org.apache.cassandra.thrift.TFramedTransportFactory; truststore=null; truststore-password=null; keystore=null; keystore-password=null; ssl-protocol=TLS; ssl-alg=SunX509; store-type=JKS; ssl-ciphers=TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA; \nPort:\n  Native Port: 9042\n  Thrift Port: 9160\n  JMX Port: 7199\nSend To Daemon:\n  *not set*\nGraph:\n  File: null\n  Revision: unknown\n  Title: null\n  Operation: WRITE\nTokenRange:\n  Wrap: false\n  Split Factor: 1\nCloudConf:\n  File: null\n\n===== Using optimized driver!!! =====\nConnected to cluster: test, max pending requests per connection null, max connections per host 8\nDatatacenter: datacenter1; Host: /127.0.31.1; Rack: rack1\nCreated keyspaces. Sleeping 1s for propagation.\nSleeping 2s...\nWarming up WRITE with 25600 iterations...\nRunning WRITE with 25 threads for 102400 iteration\ntype       total ops,    op/s,    pk/s,   row/s,    mean,     med,     .95,     .99,    .999,     max,   time,   stderr, errors,  gc: #,  max ms,  sum ms,  sdv ms,      mb\ntotal,        102400,   22221,   22221,   22221,     0.9,     0.9,     2.0,     2.0,     3.0,    17.4,    4.6,  0.00000,      0,      0,       0,       0,       0,       0\n\n\nResults:\nOp rate                   :   22,221 op/s  [WRITE: 22,221 op/s]\nPartition rate            :   22,221 pk/s  [WRITE: 22,221 pk/s]\nRow rate                  :   22,221 row/s [WRITE: 22,221 row/s]\nLatency mean              :    0.9 ms [WRITE: 0.9 ms]\nLatency median            :    0.9 ms [WRITE: 0.9 ms]\nLatency 95th percentile   :    2.0 ms [WRITE: 2.0 ms]\nLatency 99th percentile   :    2.0 ms [WRITE: 2.0 ms]\nLatency 99.9th percentile :    3.0 ms [WRITE: 3.0 ms]\nLatency max               :   17.4 ms [WRITE: 17.4 ms]\nTotal partitions          :    102,400 [WRITE: 102,400]\nTotal errors              :          0 [WRITE: 0]\nTotal GC count            : 0\nTotal GC memory           : 0.000 KiB\nTotal GC time             :    0.0 seconds\nAvg GC time               :    NaN ms\nStdDev GC time            :    0.0 ms\nTotal operation time      : 00:00:04\n\nEND\n', stderr='', rc=0)
05:58:46,719 541     commitlog_test                 DEBUG    commitlog_test.py   :899  | test_total_space_limit_of_commitlog_with_memory_based_limit: Commitlog file sizes in MB:
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/CommitLog-2-2017779.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/CommitLog-2-2017782.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/CommitLog-2-2017783.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/CommitLog-2-2017784.log
213	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/Recycled-CommitLog-2-2017786.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/Recycled-CommitLog-2-2017787.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/CommitLog-2-2017788.log
85	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/Recycled-CommitLog-2-2017789.log

05:58:46,767 541     errors                         ERROR    conftest.py         :201  | test_total_space_limit_of_commitlog_with_memory_based_limit: test failed: 
self = <commitlog_test.TestCommitLog object at 0x7f3ff42c15b0>

    def test_total_space_limit_of_commitlog_with_memory_based_limit(self):
        """
        Test with 512M commitlog files, total space limit is 3096M
        """
>       self._test_total_space_limit_of_commitlog(commitlog_segment_size_in_mb=512,
                                                  commitlog_total_space_in_mb=-1)

commitlog_test.py:944: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
commitlog_test.py:923: in _test_total_space_limit_of_commitlog
    dir_size = check_commitlog_size(allow_errors=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

allow_errors = False

    def check_commitlog_size(allow_errors: bool):
        dir_size, stdout = self._get_commitlog_size(all=True, allow_errors=allow_errors)
        if dir_size > actual_space_limit and not allow_errors:
            logger.debug(f"Commitlog file sizes in MB:\n{stdout}")
>           assert dir_size <= actual_space_limit, f"Out of total space limit\n"
E           AssertionError: Out of total space limit
E             
E           assert 808 <= 602
```

This one worries me in particular:
```
213	/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-pro65mb7/test/node1/commitlogs/Recycled-CommitLog-2-2017786.log
```

Node log is here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/149/artifact/logs-full.release.000/1669615126773_commitlog_test.py%3A%3ATestCommitLog%3A%3Atest_total_space_limit_of_commitlog_with_memory_based_limit/node1.log

**Comment by elcallio** [comment]
Does this happen repeatedly or did you only see it once? I am prone to thinking that the `total_space_limit = node1._mem_mb_per_cpu` call in the test code is not 100% equal to the `_dbcfg.available_memory / <num shards>` in actual scylla. Since the test guesses a commitlog size, but does not assign it, it seems shaky at best. But if this reproduces...
However, for me it passes. It seems to me that in the case of "default commitlog size", the test should try to pick this up from log or counters or something.

**Comment by elcallio** [comment]
The fact that on my machine, the tests idea of `total_space_limit` is 512 MB, whereas the actual `max_disk_size` computed by commitlog is 59694 MB would suggest my hunch that computed values in test are broken might have some merit...

**Comment by bhalevy** [comment]
> Does this happen repeatedly or did you only see it once?

I'm not sure about the single freak commitlog file.
But the failure are abundant in daily release test.

All 4 runs below failed:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/149/testReport/commitlog_test/TestCommitLog/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_total_space_limit_of_commitlog_with_memory_based_limit/history/

> I am prone to thinking that the total_space_limit = node1._mem_mb_per_cpu call in the test code is not 100% equal to the _dbcfg.available_memory / <num shards> in actual scylla.

The test starts scylla with a --smp=1 so `node1._mem_mb_per_cpu` should be the memory size for this signle shard.

> It seems to me that in the case of "default commitlog size", the test should try to pick this up from log or counters or something.

Feel free to propose changes to the test based on your experience and deep knowledge of how commitlog size limit works.

**Comment by elcallio** [comment]
From what I can telll, `_mem_mb_per_cpu` is always set to 512. Unless something I cannot see in ccm calls set_mem_mb_per_cpu, I don't see it changing...?

I am not sure on how to modify the test. One option is to use logs, but then we either need to promote the trace-level CL log that states how much the limit is to something more verbose, or run with trace for these tests. Is the latter a problem?

Making a counter seems like overkill. We already have too many counters... (hogging bandwidth).

**Comment by elcallio** [comment]
Btw, I do agree the segment in question is very weird. I have no explanation...

**Comment by elcallio** [comment]
That is not what I meant. I meant the log will be flooded with trace info and both very large and hard to parse in case of actual failure.

**Comment by bhalevy** [comment]
> That is not what I meant. I meant the log will be flooded with trace info and both very large and hard to parse in case of actual failure.

Hopefully here will be no failures ;-)
How large would it be?  If it's several MBs we should be fine. 100's of MB / GBs will hurt.

**Comment by bhalevy** [comment]
Anther example: https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/102/artifact/logs-full.debug.005/dtest-gw2.log
```
05:48:57,326 544     commitlog_test                 DEBUG    commitlog_test.py   :899  | test_total_space_limit_of_commitlog_with_memory_based_limit: Commitlog file sizes in MB:
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634526.log
213	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634527.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634528.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634529.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634530.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634531.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/CommitLog-2-1634533.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/Recycled-CommitLog-2-1634534.log
85	/jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.dtest/dtest-prdtrx6m/test/node1/commitlogs/Recycled-CommitLog-2-1634535.log

05:48:57,454 544     errors                         ERROR    conftest.py         :201  | test_total_space_limit_of_commitlog_with_memory_based_limit: test failed: 
self = <commitlog_test.TestCommitLog object at 0x7f2424530430>

    def test_total_space_limit_of_commitlog_with_memory_based_limit(self):
        """
        Test with 512M commitlog files, total space limit is 3096M
        """
>       self._test_total_space_limit_of_commitlog(commitlog_segment_size_in_mb=512,
                                                  commitlog_total_space_in_mb=-1)

commitlog_test.py:944: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
commitlog_test.py:923: in _test_total_space_limit_of_commitlog
    dir_size = check_commitlog_size(allow_errors=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

allow_errors = False

    def check_commitlog_size(allow_errors: bool):
        dir_size, stdout = self._get_commitlog_size(all=True, allow_errors=allow_errors)
        if dir_size > actual_space_limit and not allow_errors:
            logger.debug(f"Commitlog file sizes in MB:\n{stdout}")
>           assert dir_size <= actual_space_limit, f"Out of total space limit\n"
E           AssertionError: Out of total space limit
E             
E           assert 893 <= 602
```

**Comment by elcallio** [comment]
Question: You don't happen to have the actual commitlog segments (esp. the 213MB one) saved?

**Comment by bhalevy** [comment]
Nope, you should reproduce locally with --keep-test-dir option to keep the test nodes data directory

**Comment by elcallio** [comment]
Ran it dozens of times. No repro. :-(

**Comment by bhalevy** [comment]
> Ran it dozens of times. No repro. :-(

Did you try also with scylla in debug mode?
Maybe it might help reproduce it more easily.

**Comment by bhalevy** [comment]
Is it possible scylla doesn't take into account the recycled segments in the total limit?
For example, I reproduced the failure locally and got:
```
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:47 CommitLog-2-1230535976.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:47 CommitLog-2-1230535977.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:47 CommitLog-2-1230535978.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:48 CommitLog-2-1230535979.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:48 CommitLog-2-1230535980.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:46 CommitLog-2-1230535982.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:48 CommitLog-2-1230535983.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:46 Recycled-CommitLog-2-1230535984.log
-rwx------ 1 bhalevy bhalevy 85M Nov 30 17:47 Recycled-CommitLog-2-1230535985.log
```
The tests "actual" limit (with padding and some slack) is 602 MB,
without the recycled segments we're at 595 MB but with them we're at 765 MB.

**Comment by elcallio** [comment]
No, recycled are counted.

**Comment by bhalevy** [comment]
marking the respective dtest with `@pytest.mark.require('scylladb/scylladb#12129')`

**Comment by elcallio** [comment]
@bhaley: If you look at the log file you can see right in the beginning commitlog says its max size is 1024MB.

**Comment by bhalevy** [comment]
hmm, ok, I'll look into it.
There seems to be a disagreement of what should the memory size be in this case.

**Comment by elcallio** [comment]
Again, the node `_mem_mb_per_cpu` is wholly wrong. So the tests attempt at guessing is based on gobbletigook.

**Comment by mykaul** [comment]
> Again, the node `_mem_mb_per_cpu` is wholly wrong. So the tests attempt at guessing is based on gobbletigook.

How (and who) should set it properly?

**Comment by elcallio** [comment]
No idea. And in any case, I doubt we can guarantee it is synced with the value actually used by the internal calculation.

**Comment by bhalevy** [comment]
https://github.com/scylladb/scylla-ccm/blob/eecd64ca7cb4555616ae5602f1a57e35244be09b/ccmlib/scylla_node.py#L548-L550
is used to set the instance memory size
but I suspect that `--memory` is set in SCYLLA_EXT_OPTS....

yup, https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/102/consoleText:
```
[2022-11-30T05:46:19.529Z] SCYLLA_EXT_OPTS=--smp 2 --memory 1024M --abort-on-seastar-bad-alloc --abort-on-lsa-bad-alloc=1 --abort-on-internal-error 1
```
and so it overrides the above since the options in SCYLLA_EXT_OPTS are parsed and go into the `args` dict, here:
https://github.com/scylladb/scylla-ccm/blob/eecd64ca7cb4555616ae5602f1a57e35244be09b/ccmlib/scylla_node.py#L516-L533

Since the `--memory` option to scylla determines the total memory footprint for all shards, we should normalize the value passed in `SCYLLA_EXT_OPTS` by the `--smp` count and use the per-shard constant.

> And in any case, I doubt we can guarantee it is synced with the value actually used by the internal calculation.

@elcallio so how should the test determine the automatically calculated threshold?
Is it exposed in a metric or at least in the node log?

**Comment by elcallio** [comment]
Yes, it is in the log. I'm testing modifying the test to use parsed values instead. But there are multiple levels of wrong here...
